pkg/integrations/cloudwatch_exporter/cloudwatch_exporter.go

```diff
@@ -0,0 +1,101 @@
+package cloudwatch_exporter
+
+import (
+	"fmt"
+	"github.com/grafana/agent/pkg/integrations"
+	integrations_v2 "github.com/grafana/agent/pkg/integrations/v2"
+	"github.com/grafana/agent/pkg/integrations/v2/metricsutils"
+	"time"
+
+	"github.com/go-kit/log"
+)
+
+func init() {
+	integrations.RegisterIntegration(&Config{})
+	integrations_v2.RegisterLegacy(&Config{}, integrations_v2.TypeMultiplex, metricsutils.NewNamedShim("cloudwatch"))
+}
+
+// Config is the configuration for the CloudWatch metrics integration
+type Config struct {
+	STSRegion string          `yaml:"sts_region"`
+	Discovery DiscoveryConfig `yaml:"discovery"`
+	Static    []StaticJob     `yaml:"static"`
+}
+
+// DiscoveryConfig configures scraping jobs that will auto-discover metrics dimensions for a given service.
+type DiscoveryConfig struct {
+	ExportedTags TagsPerNamespace `yaml:"exported_tags"`
+	Jobs         []*DiscoveryJob  `yaml:"jobs"`
+}
+
+// TagsPerNamespace represents for each namespace, a list of tags that will be exported as labels in each metric.
+type TagsPerNamespace map[string][]string
+
+// DiscoveryJob configures a discovery job for a given service.
+type DiscoveryJob struct {
+	InlineRegionAndRoles `yaml:",inline"`
+	InlineCustomTags     `yaml:",inline"`
+	SearchTags           []Tag    `yaml:"search_tags"`
+	Type                 string   `yaml:"type"`
+	Metrics              []Metric `yaml:"metrics"`
+}
+
+// StaticJob will scrape metrics that match all defined dimensions.
+type StaticJob struct {
+	InlineRegionAndRoles `yaml:",inline"`
+	InlineCustomTags     `yaml:",inline"`
+	Name                 string      `yaml:"name"`
+	Namespace            string      `yaml:"namespace"`
+	Dimensions           []Dimension `yaml:"dimensions"`
+	Metrics              []Metric    `yaml:"metrics"`
+}
+
+// InlineRegionAndRoles exposes for each supported job, the AWS regions and IAM roles in which the agent should perform the
+// scrape.
+type InlineRegionAndRoles struct {
+	Regions []string `yaml:"regions"`
+	Roles   []Role   `yaml:"roles"`
+}
+
+type InlineCustomTags struct {
+	CustomTags []Tag `yaml:"custom_tags"`
+}
+
+type Role struct {
+	RoleArn    string `yaml:"role_arn"`
+	ExternalID string `yaml:"external_id"`
+}
+
+type Dimension struct {
+	Name  string `yaml:"name"`
+	Value string `yaml:"value"`
+}
+
+type Tag struct {
+	Key   string `yaml:"key"`
+	Value string `yaml:"value"`
+}
+
+type Metric struct {
+	Name       string        `yaml:"name"`
+	Statistics []string      `yaml:"statistics"`
+	Period     time.Duration `yaml:"period"`
+}
+
+// Name returns the name of the integration this config is for.
+func (c *Config) Name() string {
+	return "cloudwatch_exporter"
+}
+
+func (c *Config) InstanceKey(agentKey string) (string, error) {
+	return c.Name(), nil
+}
```

Comment:
FYI this is going to have issue running multiple instances via the v2 shim. The `InstanceKey` is used to name the instance and if more than 1 instance run (only possible via `integrations-next`) it's going to fail with duplicate name. I used parts of the config to generate a unique name with the Azure integration, https://github.com/grafana/agent/blob/aa66670de2e53d9d520f2b9c54f22823fd9d9b45/pkg/integrations/azure_exporter/azure_exporter.go#L88-L97
Since these AWS jobs can be massive not 100% sure if you can do something similar. Perhaps hashing the config?

Reply:
Yhea couldn't find a good setting, and was assuming the user might override this when using multiples instances in v2 (I know there's a config for this in the common settings). Will look into the hashing, might be a good alternative

Reply:
Used the same approach as https://github.com/grafana/agent/blob/a24f44e837c6d410623b7d88e9209e071dd29daf/pkg/metrics/instance/instance.go#L753